nightfall-client/src/services/transform.mjs

```
@@ -0,0 +1,234 @@
+      orders: commitmentInfo.leafIndices,
+      newCommitmentPreimage: commitmentInfo.newCommitments.map(o => {
+        return { value: o.preimage.value, salt: o.preimage.salt };
+      }),
+
+      recipientPublicKeys: commitmentInfo.newCommitments.map(o => o.preimage.zkpPublicKey),
+      inputTokens,
+      outputTokens,
+    };
+
+    logger.debug('computing witness...');
+    const witness = computeCircuitInputs(
+      publicData,
+      privateData,
+      commitmentInfo.roots,
+      maticAddress,
+      VK_IDS.transform.numberNullifiers,
+      VK_IDS.transform.numberCommitments,
+    );
+
+    logger.debug({
+      msg: 'witness input is',
+      witness: JSON.stringify(witness, 0, 2),
+    });
+
+    // call a worker to generate the proof
+    const res = await generateProof({ folderpath: 'transform', witness });
+
+    logger.debug({
+      msg: 'Received response from generate-proof',
+      response: JSON.stringify(res.data, null, 2),
+    });
+
+    const { proof } = res.data;
+    // and work out the ABI encoded data that the caller should sign and send to the shield contract
+
+    const optimisticTransformTransaction = new Transaction({
+      fee,
+      historicRootBlockNumberL2: commitmentInfo.blockNumberL2s,
+      circuitHash,
+      commitments: commitmentInfo.newCommitments,
+      nullifiers: commitmentInfo.nullifiers,
+      proof,
+      numberNullifiers: VK_IDS.transform.numberNullifiers,
+      numberCommitments: VK_IDS.transform.numberCommitments,
+      isOnlyL2: true,
+    });
+
+    logger.debug({
+      msg: 'Client made transaction',
+      transaction: JSON.stringify(optimisticTransformTransaction, null, 2),
+    });
+
+    return submitTransaction(
+      optimisticTransformTransaction,
+      commitmentInfo,
+      rootKey,
+      shieldContractInstance,
+      true, // offchain
+    );
+  } catch (error) {
+    await Promise.all(commitmentInfo.oldCommitments.map(o => clearPending(o)));
+    throw new Error(error);
```

You can just re-throw the error.